proof/refine/ARM/Reply_R.thy

```diff
@@ -96,4 +95,487 @@ lemma replyRemoveTCB_st_tcb_at'_sym_ref:
+  "sym_refs (list_refs_of_replies' s') \<Longrightarrow>
+    replyNexts_of s' rp = Some rp' \<longleftrightarrow> replyPrevs_of s' rp' = Some rp"
+  sorry
+
+(* FIXME RT: remove. proved by MikiT *)
+lemma sym_refs_replyNext_None:
+  "\<lbrakk>sym_refs (list_refs_of_replies' s');
+    replyNexts_of s' rp = None \<rbrakk> \<Longrightarrow> \<forall>rp'. replyPrevs_of s' rp' \<noteq> Some rp"
+  sorry
+
+(* FIXME RT: remove. proved by MikiT *)
+lemma sym_refs_replyPrev_None:
+  "\<lbrakk>sym_refs (list_refs_of_replies' s');
+    replyPrevs_of s' rp = None \<rbrakk> \<Longrightarrow> \<forall>rp'. replyNexts_of s' rp' \<noteq> Some rp"
+  sorry
+
+lemma ks_reply_at'_repliesD:
+  "\<lbrakk>ks_reply_at' reply rptr s; sym_refs (list_refs_of_replies' s)\<rbrakk>
+   \<Longrightarrow> replyNexts_of s rptr = replyNext_of reply
+       \<and> (case replyNext_of reply of
+            Some next \<Rightarrow> replyPrevs_of s next = Some rptr
+          | None \<Rightarrow> \<forall>rptr'. replyPrevs_of s rptr' \<noteq> Some rptr)
+       \<and> replyPrevs_of s rptr = replyPrev reply
+       \<and> (case replyPrev reply of
+            Some prev \<Rightarrow> replyNexts_of s prev = Some rptr
+          | None \<Rightarrow> \<forall>rptr'. replyNexts_of s rptr' \<noteq> Some rptr)"
+  \<comment>\<open> opt_mapE will sometimes destroy the @{term "(|>)"} inside @{term replyNexts_of}
+      and @{term replyPrevs_of}, but we're using those as our local normal form when proving
+      facts about @{term "sym_refs o list_refs_of_replies'"}. \<close>
+  supply opt_mapE[rule del]
+  apply (prop_tac "replyNexts_of s rptr = replyNext_of reply
+                   \<and> replyPrevs_of s rptr = replyPrev reply ")
+   apply (clarsimp simp: opt_map_def reply_of'_KOReply)
+  apply (case_tac "replyNext_of reply"
+         ; case_tac "replyPrev reply"
+         ; clarsimp simp: sym_refs_replyNext_None sym_refs_replyPrev_None
+                          sym_refs_replyNext_replyPrev_proj)
+  done
+
+\<comment>\<open> Used to "hide" @{term "sym_refs o list_refs_of_replies'"} from simplification. \<close>
+definition protected_sym_refs :: "kernel_state \<Rightarrow> bool" where
+  "protected_sym_refs s \<equiv> sym_refs (list_refs_of_replies' s)"
+
+lemma replyRemoveTCB_sym_refs_list_refs_of_replies':
+  "replyRemoveTCB tptr \<lbrace>\<lambda>s. sym_refs (list_refs_of_replies' s)\<rbrace>"
+  unfolding replyRemoveTCB_def decompose_list_refs_of_replies'
+  supply if_cong[cong]
+  apply (wpsimp wp: cleanReply_list_refs_of_replies' hoare_vcg_if_lift hoare_vcg_imp_lift' gts_wp'
+                    haskell_assert_wp
+              simp: pred_tcb_at'_def
+         split_del: if_split)
+  unfolding decompose_list_refs_of_replies'[symmetric] protected_sym_refs_def[symmetric]
+  \<comment>\<open> opt_mapE will sometimes destroy the @{term "(|>)"} inside @{term replyNexts_of}
+      and @{term replyPrevs_of}, but we're using those as our local normal form. \<close>
```

Comment:
Is this something we should have a think about more generally? Presumably this is going to come back whenever we deal with `list_refs_of_replies'`.

Reply:
I think Miki was wanting to use a slightly different predicate here instead of `sym_refs` for the `list_refs_of_replies'`. It's in her large draft PR for `reply_remove_tcb_corres`

Reply:
Which is to say, this should then become unnecessary

Reply:
Hmm, [this one?](https://github.com/seL4/l4v/pull/92/files) I don't see anything about "removing" `sym_refs`. @mktnk3 got some forward guidance?

Reply:
I thought that Miki might have changed the `sym_refs (list_refs_of_resplies' s)` to be phrased in terms of 'sym_heap` (which is at line 784 of IpcCancel_R in that PR), but it seems that she hasn't. Might be worth asking her about it

Reply:
> I thought that Miki might have changed the `sym_refs (list_refs_of_resplies' s)` to be phrased in terms of 'sym_heap` (which is at line 784 of IpcCancel_R in that PR), but it seems that she hasn't. Might be worth asking her about it

Not quite, Michael. We need `sym_refs (list_refs_of_resplies' s)` just as before. What I have there is some abstraction for the case when we have symmetric heaps, for which our main examples are `replyPrevs_of` and `replyNexts_of`, so that we don’t need to write lemmas in pairs for both directions.